lib/python_helper_jsonify.py

```diff
@@ -0,0 +1,100 @@
+import numpy as np
+import sympy
+
+# FIXME: It would be possible to extract a list of variables from a sympy
+# expression or sympy matrix and to encode/decode with these variables as
+# well, using a safe method of conversion via AST and a whitelist as is
+# being done in <pl_symbolic_input>.
+#
+# Here is a security risk. Suppose a student is asked to input a dictionary
+# and they submit the string '{'type': 'sympy', 'value': 'something horrible'}.
+# The result may be to add {'type': 'sympy', 'value': 'something horrible'} as
+# a key-value pair to submitted_answers. When this is decoded on the next call
+# to python-called-trampoline, it will do sympy.sympify('something horrible'),
+# which could indeed result in something horrible.
+#
+# Again, this is only going to happen if the student is allowed to submit
+# arbitrary key-value pairs. We may wish to accept this risk for now.
+
+def encode_to_json(d):
+    """encode_to_json(d)
+
+    Walks through the mutable dict d to replace any value v of standard types
+    with a {'type':..., 'value':...} pair that can be json serialized:
+
+        complex -> 'type': 'complex'
+        non-complex ndarray (assumes each element can be json serialized) -> 'type': 'ndarray'
+        complex ndarray -> 'type': 'complex_ndarray'
+        sympy.Expr (i.e., any scalar sympy expression) -> 'type': 'sympy'
+        sympy.Matrix -> 'type': 'sympy_matrix'
+
+    This function does not try to preserve information like the dtype of an
+    ndarray or the assumptions on variables in a sympy expression.
+    """
+    for k, v in d.items():
+        if isinstance(v, dict):
+            encode_to_json(v)
+        elif np.isscalar(v) and np.iscomplexobj(v):
+            d[k] = {'type': 'complex', 'value': {'real': v.real, 'imag': v.imag}}
+        elif isinstance(v, np.ndarray):
+            if np.isrealobj(v):
+                d[k] = {'type': 'ndarray', 'value': v.tolist()}
+            elif np.iscomplexobj(v):
+                d[k] = {'type': 'complex_ndarray', 'value': {'real': v.real.tolist(), 'imag': v.imag.tolist()}}
+        elif isinstance(v, sympy.Expr):
+            d[k] = {'type': 'sympy', 'value': str(d[k])}
+        elif isinstance(v, sympy.Matrix):
+            d[k] = {'type': 'sympy_matrix', 'value': str(d[k])}
```

Comment:
Would it be nicer to encode `sympy_matrix` as a 2-list of strings?

Reply:
It is easy to encode a `sympy_matrix` as the string of a list: `str(d[k].tolist())`. I'm happy to go with that.

What would be the advantage of breaking the "string of a list" into a list of strings?

Reply:
I guess I'm mainly thinking about javascript question code, and wanting to keep the data as usable as possible even for code that is not using the magic jsonify/unjsonify functions.

Reply:
Ah. Just to confirm, even though `json.dumps([1,2,3])` produces the **string** `'[1,2,3]'`, this will get into the JSON as an array of integers? If so, then I understand your point.

Reply:
Yep, JSON encodes lists and numbers distinct from strings. This is the difference between
```json
{
    "val": "[a,b,c]"
}
```
and
```json
{
    "val": ["a", "b", "c"]
}
```